design/ceph/cephx-keyring-rotation.md

```diff
@@ -0,0 +1,353 @@
+
+Rook operator configuration option `ROOK_ROTATE_LOCAL_CEPHX_KEYS_OLDER_THAN` -- allowed values:
+
+- `CurrentCephVersion`: rotate local CephX keys when the Ceph version is updated
+- **else**: do not rotate local CephX keys
+
+Rook could use a boolean true/false value for this config instead of an "older-than" selection, but
+other rotation configs (below) rely on "older-than" interfaces, and consistency will help users and
+will also allow flexibility for the future of the config in the face of unforeseen needs.
+
+List of Ceph daemons which have local keys that can be rotated automatically:
+
+- mon - all share the same key
+- mgr
+- osd
+- mds
+- rgw
+- nfs
+- rbdmirror - has 2 keys: one local (will auto-rotate) and one mirror peer (will not auto-rotate)
+- cephfsmirror - same considerations as rbdmirror
+- log-rotator
+- crash-collector
+
+The Ceph admin key Rook relies on to perform Ceph admin will also be rotated automatically. Care
+should be taken to ensure admin key rotation doesn't block rotation of other keys.
+
+### Enabling CSI key rotation
+
+Ceph-CSI resources (if managed by Rook) are managed in the operator namespace, so an operator
+configuration option is most appropriate. CSI key rotations may require manual administrator action
+to reboot or drain/undrain nodes to PVCs, so Rook should allow users to initiate one-time key
+rotations.
+
+Rook operator configuration option `ROOK_ROTATE_CSI_CEPHX_KEYS_OLDER_THAN` -- allowed values:
+
+- `<DATETIME>`: rotate local CephX keys that were minted prior to the input DATETIME string.
+    e.g., `Thu May 29 18:08:03 UTC 2025` (from `date -u` command)
+- **else**: do not rotate CSI CephX keys
+
+While usage of a DATETIME string isn't ideal, I believe that it provides a usable UX for initiating
+one-time rotations.
+
+Rook could use a string like `once` as an input, but Rook would have to record when `once` was first
+observed so that it doesn't repeat rotations with every reconcile.
+
+Rook could alternatively use a string like `always` and expect the user to unset `always` after they
+see that rotation is complete, but this requires the user to monitor the rotation status and change
+config in a clunky way that might risk manual errors.
+
+Rook could use the Ceph version as the "older-than" selection, but that would only allow CSI key
+rotations when the Ceph version changes. For a user already at the latest Ceph version, they
+wouldn't have the option to rotate CSI keys on demand.
+
+### Enabling non-local key rotation
+
+Ceph daemons that have keys used by non-Rook-controlled clients are also associated with
+Custom Resources (CRs).
+
+Full list of Rook CRs with non-local keys:
+
+- CephClient - the created arbitrarily-usable client has a single associated key
+- CephRBDMirror - mirror daemon has a mirror peer key
+- CephFilesystemMirror - mirror daemon has a mirror peer key
+
+Each CR will provide a key rotation mechanism as part of the primary API spec.
```

Comment:
Why do you need this for each CR? Shouldn't this more like a global API that will rotate the keys for all the resources? I don't we customer would just want to rotate the keys for RDBMirror and not FSMirror.